packages-exp/functions-exp/src/service.ts

```diff
@@ -0,0 +1,269 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FirebaseApp } from '@firebase/app-types-exp';
+import {
+  HttpsCallable,
+  HttpsCallableResult,
+  HttpsCallableOptions
+} from '@firebase/functions-types-exp';
+import { _errorForResponse, HttpsErrorImpl } from './error';
+import { ContextProvider } from './context';
+import { Serializer } from './serializer';
+import { Provider } from '@firebase/component';
+import { FirebaseAuthInternalName } from '@firebase/auth-interop-types';
+import { FirebaseMessagingName } from '@firebase/messaging-types';
+
+export const DEFAULT_REGION = 'us-central1';
+
+/**
+ * The response to an http request.
+ */
+interface HttpResponse {
+  status: number;
+  json: HttpResponseBody | null;
+}
+/**
+ * Describes the shape of the HttpResponse body.
+ * It makes functions that would otherwise take {} able to access the
+ * possible elements in the body more easily
+ */
+export interface HttpResponseBody {
+  data?: unknown;
+  result?: unknown;
+  error?: {
+    message?: unknown;
+    status?: unknown;
+    details?: unknown;
+  };
+}
+
+/**
+ * Returns a Promise that will be rejected after the given duration.
+ * The error will be of type HttpsErrorImpl.
+ *
+ * @param millis Number of milliseconds to wait before rejecting.
+ */
+function failAfter(millis: number): Promise<never> {
+  return new Promise((_, reject) => {
+    setTimeout(() => {
+      reject(new HttpsErrorImpl('deadline-exceeded', 'deadline-exceeded'));
+    }, millis);
+  });
+}
+
+/**
+ * The main class for the Firebase Functions SDK.
+ * @internal
+ */
+export class FunctionsService {
+  readonly contextProvider: ContextProvider;
+  readonly serializer = new Serializer();
```

Comment:
I think serializer is stateless? If so, we can create a single instance outside the class and let all functions instances share it.

Reply:
Sure, I actually am not even sure why it is a class?

Reply:
Good point. We can just have 2 functions `encode()` and `decode()` instead.

Reply:
Done.